Stack/Opc.Ua.Core/Stack/Tcp/TcpMessageSocket.cs

```diff
@@ -690,6 +677,88 @@ private void ReadNextBlock()
                 throw ServiceResultException.Create(StatusCodes.BadTcpInternalError, ex, "BeginReceive failed.");
             }
         }
+
+        /// <summary>
+        /// delegate to handle internal callbacks with socket error
+        /// </summary>
+        private delegate void CallbackAction(SocketError error);
+
+        /// <summary>
+        /// Try to connect to endpoint and do callback if connected successfully
+        /// </summary>
+        /// <param name="address">Endpoint address</param>
+        /// <param name="addressFamily">Endpoint address family</param>
+        /// <param name="port">Endpoint port</param>
+        /// <param name="callback">Callback that must be executed if the connection would be established</param>
+        private SocketError BeginConnect(IPAddress address, AddressFamily addressFamily, int port, CallbackAction callback)
+        {
+            var socket = new Socket(addressFamily, SocketType.Stream, ProtocolType.Tcp);
+            var args = new SocketAsyncEventArgs()
+            {
+                UserToken = callback,
+                RemoteEndPoint = new IPEndPoint(address, port),
+            };
+            args.Completed += new EventHandler<SocketAsyncEventArgs>(OnSocketConnected);
+            lock (m_socketLock)
+            {
+                m_socketResponses++;
+            }
+            if (!socket.ConnectAsync(args))
+            {
+                // I/O completed synchronously
+                OnSocketConnected(socket, args);
+                return args.SocketError;
+            }
+            return SocketError.InProgress;
+        }
+
+        /// <summary>
+        /// Handle socket connection event
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        private void OnSocketConnected(object sender, SocketAsyncEventArgs args)
+        {
+            var socket = sender as Socket;
+            bool success = false;
+            lock (m_socketLock)
+            {
+                m_socketResponses--;
+                if (!m_closed && m_socket == null)
+                {
+                    if (args.SocketError == SocketError.Success)
+                    {
+                        m_socket = socket;
+                        success = true;
+                        m_tcs.SetResult(args.SocketError);
+                    }
+                    else if (m_socketResponses == 0)
```

Comment:
Could we hit a condition where not all ip addresses are checked, but the timing between currently tested ipv4 and ipv6 addresses is that we hit m_socketResponses == 0? Would it not make more sense to intitialize this variable to the max number of addresses to test?

Reply:
The socketResponses counter increases, e.g. if there are timeouts on connectionattempts (typ. 15s) we try new Ipv4/Ipv6 address pairs after every second. Then the last response will cause the error return. Fast ICMP responses, e.g. if the server port is not right (ConnectionRefused) we hit the zero count pretty fast but still a new IPv4/6 address pair is needed...